### PR TITLE
Finagle scps to bypass aws limit 

### PR DESCRIPTION
**Pull request by mdawn**
### [New module! ou-scp to combine SCPs](https://www.pivotaltracker.com/story/show/173455007)

Relevant Info:
- You can use this branch to test with: [mdawn-test-out-module-ou-scp-173455007](https://github.com/trussworks/legendary-waddle/pull/new/mdawn-test-out-module-ou-scp-173455007)
- PR to add it to our GH repos:
https://github.com/trussworks/legendary-waddle/pull/266
- Release `v1.0.0` is in draft, to be published after merge
- Subscribed to in [`#infra-feed`](https://trussworks.slack.com/archives/C91SHMKFV/p1593460432258400)

Does these things:
- makes a new repo based on [the module template](https://github.com/trussworks/terraform-module-template)
- gives option to combine statements from pre-existing policies in [`terraform-aws-org-scp`]( https://github.com/trussworks/terraform-aws-org-scp) 
- adds locals & vars to toggle between adding either a hard-coded "Deny" policy statement or an empty array of strings
- adds sids to each statement 
- `generated policy` switches between `deny_all` and combined policy
Based loosely on the structure used in [`terraform-aws-logs`](https://github.com/trussworks/terraform-aws-logs/blob/master/main.tf)

---
sample module call for all policies in a combined policy block (see readme for more options):

```HCL
module "github_terraform_aws_ou_scp" {
  source = "trussworks/ou-scp/aws"
  target =  aws_organizations_organizational_unit.my_ou

  # don't allow all accounts to be able to leave the org
  deny_leaving_orgs             = true
  # applies to accounts that are not managing IAM users
  deny_creating_iam_users       = true
  # don't allow deleting KMS keys
  deny_deleting_kms_keys        = true
  # don't allow deleting Route53 zones
  deny_deleting_route53_zones   = true
  # don't allow deleting CloudWatch logs
  deny_deleting_cloudwatch_logs = true
  # don't allow access to the root user
  deny_root_account             = true

  protect_s3_buckets            = true
  # protect terraform statefile bucket
  protect_s3_bucket_resources   = [
    "arn:aws:s3:::prod-terraform-state-us-west-2",
    "arn:aws:s3:::prod-terraform-state-us-west-2/*"
  ]

  protect_iam_roles             = true
  # - protect OrganizationAccountAccessRole
  protect_iam_role_resources     = [
    "arn:aws:iam::*:role/OrganizationAccountAccessRole"
  ]

  # restrict region-specific operations to us-west-2
  limit_regions                 = true
  # - restrict region-specific operations to us-west-2
  allowed_regions               = ["us-west-2"]

  # require s3 objects be encrypted
  require_s3_encryption = true
}
```

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1 +1,231 @@
...
+  #
+  # Deny all access
+  #
+
+  statement {
+    sid       = "DenyAllAccess"
+    effect    = local.deny_all_access_effect
+    actions   = ["*"]
+    resources = ["*"]
+  }
```
Comment (cblkwell) [inline main.tf:87]: We talked about this earlier -- this should be in a separate policy document, and then we want a switch to use either the one with all the other policies, or this one.

File: variables.tf [hunk 2]
```diff
@@ -1 +1,94 @@
...
+  type        = list(string)
+  default     = [""]
+}
+
+variable "target" {
+  description = "id of OU or account on which to attach SCP"
+  type        = string
+}
+
+variable "add_deny_root_account" {
```
Comment (cblkwell) [inline variables.tf:24]: I do not like the name of these variables -- the `add` part adds nothing, and I think it makes things more confusing. Just drop it from everything.

File: variables.tf [hunk 3]
```diff
@@ -1 +1,94 @@
...
+  description = "Add LimitRegions to the OU policy."
+  default     = false
+  type        = bool
+}
+
+# variable "add_all_policies" {
+#   description = "Add all policies in this module to the OU policy."
+#   default     = false
+#   type        = bool
+# }
```
Comment (cblkwell) [inline variables.tf:94]: I'd recommend just removing this. The cases where this would be set are miniscule.

File: variables.tf (hunk 2 above)
Comment (mdawn) [reply variables.tf:24]: yeah this was just so I could get my head around `allow_deny`. I realize it's less than ideal.

File: main.tf [hunk 4]
```diff
@@ -1 +1,240 @@
...
+  deny_leaving_orgs_effect             = var.deny_leaving_orgs ? "Deny" : "Allow"
+  deny_creating_iam_users_effect       = var.deny_creating_iam_users ? "Deny" : "Allow"
+  deny_deleting_kms_keys_effect        = var.deny_deleting_kms_keys ? "Deny" : "Allow"
+  deny_deleting_route53_zones_effect   = var.deny_deleting_route53_zones ? "Deny" : "Allow"
+  require_s3_encryption_effect         = var.require_s3_encryption ? "Deny" : "Allow"
+  deny_deleting_cloudwatch_logs_effect = var.deny_deleting_cloudwatch_logs ? "Deny" : "Allow"
+  protect_s3_buckets_effect            = var.protect_s3_buckets ? "Deny" : "Allow"
+  protect_iam_roles_effect             = var.protect_iam_roles ? "Deny" : "Allow"
+  limit_regions_effect                 = var.limit_regions ? "Deny" : "Allow"
+}
```
Comment (cblkwell) [inline main.tf:213]: This seems like a weird place for these declarations -- maybe put them at the beginning of the file?

File: variables.tf [hunk 5]
```diff
@@ -1 +1,90 @@
...
+variable "allowed_regions" {
+  description = "AWS Regions allowed for use (for use with the restrict regions SCP)"
+  type        = list(string)
+  default     = [""]
+}
+
+variable "target" {
+  description = "id of OU or account on which to attach SCP"
+  type        = string
+}
```
Comment (cblkwell) [inline variables.tf:22]: I'm going to suggest you change this to an `object` type variable and have it take the actual OU resource, not the ID of the OU. This will let you get the name of the OU for naming the SCP; so you'd change this to:
```suggestion
variable "target" {
  description = "OU resource to attach SCP"
  type        = object({
    name = string
    id = string
  })
}
```

File: main.tf [hunk 6]
```diff
@@ -1 +1,240 @@
...
+  name        = "deny-all-access"
+  description = "Deny all access"
+  content     = data.aws_iam_policy_document.deny_all_access.json
+}
+
+resource "aws_organizations_policy_attachment" "deny_all_access" {
+  count     = var.create_combined_policy_block ? 0 : 1
+  policy_id = aws_organizations_policy.deny_all_access[0].id
+  target_id = var.target
+}
```
Comment (cblkwell) [inline main.tf:240]: Ugh, evidently my comment on this vanished -- your second policy declaration here is not necessary, you're actually going to run into some problems because you're giving these policies a set name, but it's very likely you're going to be applying this module more than once, which will create collisions. Instead, I suggest replacing both `aws_organizations_policy` and `aws_organizations_policy_attachment` resources with something like this:

```hcl
resource "aws_organizations_policy" "generated" {
  name = "${var.target.name}-generated-ou-scp"
  description = "${var.target.name} SCP generated by ou-scp module"
  content = var.deny_all ? data.aws_iam_policy_document.deny_all_access.json : data.aws_iam_policy_document.combined_policy_block.json
}

resource "aws_organizations_policy_attachment" "generated" {
  policy_id = aws_organizations_policy.generated.id
  target_id = var.target.id
}
```

File: variables.tf [hunk 7]
```diff
@@ -1 +1,90 @@
...
+
+# Policy Switches
+
+variable "create_combined_policy_block" {
```
[the remark this answers is not in the document]
Comment (mdawn) [reply variables.tf:26]: Yeah good suggestion. I just followed the `create_public_access_block` pattern in `aws-logs` here & didn't think much about the name when adding the switch.

File: README.md [hunk 8]
```diff
@@ -23,15 +25,37 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
+  deny_root_account = true
+  deny_leaving_orgs = true
+  deny_creating_iam_users = true
+  deny_deleting_kms_keys = true
+  deny_deleting_route53_zones = true
+  require_s3_encryption = true
+  deny_deleting_cloudwatch_logs = true
+  protect_s3_buckets = true
+  protect_iam_roles = true
+  limit_regions = true
```
Comment (cblkwell) [inline README.md:46]: If you have "true" set you need to provide some regions to limit it too, same for the IAM and S3 bucket protections. If you're going to have an example like this you should add those.

File: README.md [hunk 9]
```diff
@@ -23,15 +25,37 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
 Terraform 0.11. Pin module version to ~> 1.0. Submit pull-requests to terraform011 branch.
 
-## Usage
+### Usage for combined policy statements
+
+```hcl
+// TODO: make sure these references to module names are correct once pushed to the registry
+module "ou_scp" {
+  source = "trussworks/ou-scp"
+  target = aws_organizations_organizational_unit.root.id
```
Comment (cblkwell) [inline README.md:34]: For your example, I'd use a more hypothetical OU like `aws_organizations_organizational_unit.my_ou.id` or something, just to keep people from doing something weird on accident.

File: README.md [hunk 10]
```diff
@@ -23,15 +25,37 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
-### Put an example usage of the module here
+### Usage for a policy which denies all access
 
 ```hcl
-module "example" {
-  source = "terraform/registry/path"
+module "scp_test_scp" {
+  # source needs to be the most recent commit hash
+  source = "trussworks/ou-scp"
+  target = aws_organizations_organizational_unit.root.id
```
Comment (cblkwell) [inline README.md:56]: Same comment as above.

File: examples/simple/main.tf [hunk 11]
```diff
@@ -1,3 +1,8 @@
 # ---------------------------------------------------------------------------------------------------------------------
 # BASIC TERRAFORM CONTENT EXAMPLE
 # ---------------------------------------------------------------------------------------------------------------------
+
+module "aws_ou_scp" {
+  # default_allow = false
+  add_deny_root_account = true
+}
```
Comment (cblkwell) [inline examples/simple/main.tf:8]: You might just want to get rid of this. Automated testing for organization modules is pretty hard, since it requires doing something to an actual organization, and most people don't have more than one.

File: README.md (hunk 8 above)
Comment (mdawn) [reply README.md]: 😬  idk how I missed those...

File: README.md [hunk 12]
```diff
@@ -23,15 +25,58 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
 ```hcl
-module "example" {
-  source = "terraform/registry/path"
+// TODO: make sure these references to module names are correct once pushed to the registry
+module "ou_scp" {
+  source = "trussworks/ou-scp"
+  target = {
+    name = "my_ou"
+    id = aws_organizations_organizational_unit.my_ou.id
+  }
```
Comment (cblkwell) [inline README.md:37]: This should just look like this:
```suggestion
  target = aws_organizations_organizational_unit.my_ou
```

Your target is the actual OU resource, the variable definition just pulls those attributes we added in `variables.tf` from the object directly.

File: README.md [hunk 13]
```diff
@@ -23,15 +25,58 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
 
-  <variables>
+```hcl
+module "scp_test_scp" {
+  # source needs to be the most recent commit hash
+  source = "trussworks/ou-scp"
+  target = {
+    name = "my_ou"
+    id = aws_organizations_organizational_unit.my_ou.id
+  }
```
Comment (cblkwell) [inline README.md:77]: Same thing as above.

File: main.tf [hunk 14]
```diff
@@ -1 +1,234 @@
...
+  deny_leaving_orgs_effect             = var.deny_leaving_orgs ? "Deny" : "Allow"
+  deny_creating_iam_users_effect       = var.deny_creating_iam_users ? "Deny" : "Allow"
+  deny_deleting_kms_keys_effect        = var.deny_deleting_kms_keys ? "Deny" : "Allow"
+  deny_deleting_route53_zones_effect   = var.deny_deleting_route53_zones ? "Deny" : "Allow"
+  deny_deleting_cloudwatch_logs_effect = var.deny_deleting_cloudwatch_logs ? "Deny" : "Allow"
+  deny_root_account_effect             = var.deny_root_account ? ["Deny"] : []
+  protect_s3_buckets_effect            = var.protect_s3_buckets ? ["Deny"] : []
+  protect_iam_roles_effect             = var.protect_iam_roles ? ["Deny"] : []
+  limit_regions_effect                 = var.limit_regions ? ["Deny"] : []
+}
```
Comment (cblkwell) [inline main.tf:11]: Why are these lists?

File: variables.tf [hunk 15]
```diff
@@ -1 +1,89 @@
...
+
+variable "deny_root_account" {
+  description = "DenyRootAccount in the OU policy."
+  default     = false
+  type        = bool
+}
+
+variable "protect_s3_buckets" {
+  description = "ProtectS3Buckets in the OU policy."
+  default     = true
```
Comment (cblkwell) [inline variables.tf:75]: Why is this defaulting to true?

File: main.tf (hunk 14 above)
Comment (mdawn) [reply main.tf:11]: Funny story: AWS only acknowledges some of these policies as "Deny". When I tried to set the `effect` to either Deny or Allow, I get a malformed policy document error on "Allow". Using `dynamic` + `for_each` and then setting the `effect` to either "Deny" or an empty array prevents the "Allow" problem and lets the `effect` (and related variable switch) work. 

If it's stupid but it works...it ain't stupid?

File: variables.tf (hunk 15 above)
Comment (mdawn) [reply variables.tf:75]: Good question 😆 - thanks

File: README.md (hunk 13 above)
Comment (mdawn) [reply README.md:83]: Thanks. I could've sworn I fixed this already.

File: main.tf (hunk 14 above)
Comment (cblkwell) [reply main.tf:11]: I guess I'm just not sure how a list is being interpreted in a field that expects either "Allow" or "Deny". That seems...weird.

Comment (mdawn) [reply main.tf:11]: It is def a bit weird.

The `effect` field is only ever "Deny" for these. We're either including the "Deny" statement (with hard-coded Deny in the `content` param ) or passing no statement through at all. For naming, the Deny/(alternative) pattern just provides some consistency in the module call.

File: main.tf [hunk 16]
```diff
@@ -1,13 +1,13 @@
...
+  deny_leaving_orgs_statement             = var.deny_leaving_orgs ? [""] : []
+  deny_creating_iam_users_statement       = var.deny_creating_iam_users ? [""] : []
+  deny_deleting_kms_keys_statement        = var.deny_deleting_kms_keys ? [""] : []
+  deny_deleting_route53_zones_statement   = var.deny_deleting_route53_zones ? [""] : []
+  deny_deleting_cloudwatch_logs_statement = var.deny_deleting_cloudwatch_logs ? [""] : []
+  deny_root_account_statement             = var.deny_root_account ? [""] : []
+  protect_s3_buckets_statement            = var.protect_s3_buckets ? [""] : []
+  protect_iam_roles_statement             = var.protect_iam_roles ? [""] : []
+  limit_regions_statement                 = var.limit_regions ? [""] : []
 }
```
Comment (cblkwell) [inline main.tf]: Please add a comment explaining how this works with the `dynamic` directives so this is a little more clear.

File: outputs.tf [hunk 17]
```diff
@@ -1 +1,4 @@
 # Outputs placeholder
+output "scp_policy_json" {
+  value = data.aws_iam_policy_document.combined_policy_block.json
+}
```
Comment (cblkwell) [inline outputs.tf:4]: Why are you adding this as an output?

Comment (mdawn) [reply outputs.tf:4]: Good eye. It was helping me debug / see if the malformed policy error was literally a json problem or not. I'll take it out.

File: main.tf (hunk 16 above)
Comment (mdawn) [reply main.tf]: I did this but it might be a bit wordy. Lmk if you see an obvious way to pare it down.

File: README.md [hunk 18]
```diff
@@ -23,15 +26,53 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
-## Usage
+### Usage for combined policy statements
 
-### Put an example usage of the module here
+To include a policy in your combined policy block, set it to `true`. Otherwise omit the policy variable.
 
 ```hcl
-module "example" {
-  source = "terraform/registry/path"
+// TODO: make sure these references to module names are correct once pushed to the registry
```
Comment (chrisgilmerproj) [inline README.md:34]: TODO

File: README.md [hunk 19]
```diff
@@ -23,15 +26,53 @@ Terraform 0.12. Pin module version to ~> 2.0. Submit pull-requests to master bra
...
+module "ou_scp" {
+  source = "trussworks/ou-scp"
```
Comment (chrisgilmerproj) [inline README.md:36]:
```suggestion
  source = "trussworks/ou-scp/aws"
```

My experience is that the naming convention of the repo maps to the source module name.

File: go.mod [hunk 20]
```diff
@@ -1,15 +1,3 @@
-module github.com/trussworks/terraform-module-template
+module terraform-aws-ou-scp
```
Comment (chrisgilmerproj) [inline go.mod:1]: Can this file just be deleted?

File: README.md (hunk 19 above)
Comment (eeeady) [reply README.md:36]: thought it would be ehere.... https://github.com/trussworks/Engineering-Playbook/tree/master/infrasec/tf#going-public but I think Chris is right.

File: go.mod (hunk 20 above)
Comment (eeeady) [reply go.mod:1]: Think so... there aren't terratests to support with this.

File: main.tf [hunk 21]
```diff
@@ -1 +1,267 @@
...
+  deny_leaving_orgs_statement                = var.deny_leaving_orgs ? [""] : []
+  deny_creating_iam_users_statement          = var.deny_creating_iam_users ? [""] : []
+  deny_deleting_kms_keys_statement           = var.deny_deleting_kms_keys ? [""] : []
+  deny_deleting_route53_zones_statement      = var.deny_deleting_route53_zones ? [""] : []
+  deny_deleting_cloudwatch_logs_statement    = var.deny_deleting_cloudwatch_logs ? [""] : []
+  deny_root_account_statement                = var.deny_root_account ? [""] : []
+  protect_s3_buckets_statement               = var.protect_s3_buckets ? [""] : []
+  protect_iam_roles_statement                = var.protect_iam_roles ? [""] : []
+  limit_regions_statement                    = var.limit_regions ? [""] : []
+  deny_unencrypted_object_uploads_statement  = var.require_s3_encryption ? [""] : []
```
Comment (eeeady) [inline main.tf:19]: NICEEEE!

File: README.md (hunk 19 above)
Comment (cblkwell) [reply README.md:36]: Chris is definitely correct. You can look at any of the other module calls in our Terraform codebase.

Comment (mdawn) [reply README.md:36]: I'm getting the vague sense that Chris might be right about this...
